File: src/main/java/com/yerti/core/recipe/CustomRecipeHandler.java
package com.yerti.core.recipe;

import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.inventory.PrepareItemCraftEvent;
import org.bukkit.inventory.CraftingInventory;
import org.bukkit.inventory.ItemStack;

import java.util.Map;

public class CustomRecipeHandler implements Listener {


    @EventHandler
    public void onPrepareCraft(PrepareItemCraftEvent event) {

        CraftingInventory inventory = event.getInventory();

        if (event.getInventory().getResult() == null) return;
        if (CustomRecipe.getMatrixMap() == null) return;

        for (Map.Entry<ItemStack, ItemStack[]> entry : CustomRecipe.getMatrixMap().entrySet()) {


            if (inventory.getResult().equals(entry.getKey())) {

                for (int i = 0; i < 9; i++) {

                    if (inventory.getMatrix()[i].hasItemMeta()) {
                        if (!entry.getValue()[i].hasItemMeta()) {
                            return;
                        }
                    }


                    if (!inventory.getMatrix()[i].getItemMeta().equals(entry.getValue()[i].getItemMeta())) {
                        inventory.setResult(null);
                        return;
                    }
                }


            }

        }


    }

}

File: src/main/java/com/yerti/core/recipe/CustomRecipeHandler.kt
package com.yerti.core.recipe

import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.inventory.PrepareItemCraftEvent

class CustomRecipeHandler : Listener {
    @EventHandler
    fun onPrepareCraft(event: PrepareItemCraftEvent) {
        val inventory = event.inventory
        if (event.inventory.result == null) return
        if (CustomRecipe.Companion.getMatrixMap() == null) return
        for ((key, value) in CustomRecipe.Companion.getMatrixMap()!!.entries) {
            if (inventory.result == key) {
                for (i in 0..8) {
                    if (inventory.matrix[i].hasItemMeta()) {
                        if (!value[i]!!.hasItemMeta()) {
                            return
                        }
                    }
                    if (inventory.matrix[i].itemMeta != value.get(i)!!.itemMeta) {
                        inventory.result = null
                        return
                    }
                }
            }
        }
    }
}